### PR TITLE
mixpool: Cache recently removed msgs.

### PR DESCRIPTION
**Pull request by davecgh**
**This requires #3360 and is rebased on #3362**.

This adds a cache to house mix messages that have recently been removed from the mixpool.  It makes use of the new `container/lru` module to handle automatic expiration of entries and maximum entry limiting.

The rationale for this change is that it is considered misbehavior to advertise a mix message and then claim it is not found when the corresponding request arrives.  Maintaining a separate cache of mix messages recently removed from the mixpool for a short period of time significantly increases the probability they are available to serve when a request for the advertisement arrives independent of the current status of the mixpool.

### CODE REVIEW

File: mixing/mixpool/mixpool.go [hunk 1]
```diff
@@ -236,6 +282,21 @@ func (p *Pool) HaveMessage(query *chainhash.Hash) bool {
 	return ok
 }
 
+// RecentMessage searches the mixing pool and cache of recent messages that have
+// been removed for a message by its hash.
```
Comment (jrick) [inline mixing/mixpool/mixpool.go:286]: i would reword this comment. it is not searching for only recent messages that have been removed, but any messages that is currently accepted plus any message that was recently removed and still contained in the lru.

Comment (jrick) [reply mixing/mixpool/mixpool.go:286]: maybe (well, definitely) it's just because english is dumb and i could see this being interperted multiple ways, but it's definitely not clearly written in any case.

**Review by jrick — APPROVED** [review]
code-wise with the exception of the nitpick i had on the comment wording, this looks ok.

File: mixing/mixpool/mixpool.go (hunk 1 above)
Comment (davecgh) [reply mixing/mixpool/mixpool.go:286]: Yeah, I meant it as in that is searches the mixing pool and (then if not found) the cache of recent messages, but I can definitely see what you mean that it's easy to read it the other way.

I agree your wording is clearer.  How about this?

> RecentMessages attempts to find a message by its hash in both the mixing pool that contains accepted messages as well as the cache of recently removed messages.

Comment (jrick) [reply mixing/mixpool/mixpool.go:286]: sure, def better

Comment (davecgh) [reply mixing/mixpool/mixpool.go]: Updated.
